### PR TITLE
cardano-ledger-specs annotated decoding

### PR DESCRIPTION
**Pull request by redxaxder**

### CODE REVIEW

File: ouroboros-consensus-shelley/src/Ouroboros/Consensus/Shelley/Ledger/Block.hs [hunk 1]
```diff
@@ -182,8 +183,8 @@ instance Crypto c => ToCBOR (ShelleyBlock c) where
   -- Don't encode the header hash, we recompute it during deserialisation
   toCBOR = toCBOR . shelleyBlockRaw
```
[the remark this answers is not in the document]
Comment (redxaxder) [reply ouroboros-consensus-shelley/src/Ouroboros/Consensus/Shelley/Ledger/Block.hs:184]: Yes. This is in the ToCBOR instance.

File: ouroboros-consensus-shelley/src/Ouroboros/Consensus/Shelley/Node.hs [hunk 2]
```diff
@@ -58,6 +58,7 @@ import qualified Shelley.Spec.Ledger.STS.Chain as SL
 import qualified Shelley.Spec.Ledger.STS.Prtcl as SL
 import qualified Shelley.Spec.Ledger.TxData as SL
 import qualified Shelley.Spec.Ledger.UTxO as SL
+import qualified Shelley.Spec.Ledger.Serialization as SL
```
Comment (intricate) [inline ouroboros-consensus-shelley/src/Ouroboros/Consensus/Shelley/Node.hs:61]: This should come after the `Shelley.Spec.Ledger.PParams` import.

File: ouroboros-consensus-shelley/test/Test/Consensus/Shelley/Ledger.hs [hunk 3]
```diff
@@ -46,6 +46,8 @@ import qualified Shelley.Spec.Ledger.LedgerState as SL
 import qualified Shelley.Spec.Ledger.OCert as SL
 import qualified Shelley.Spec.Ledger.PParams as SL
 import qualified Shelley.Spec.Ledger.Rewards as SL
+import qualified Shelley.Spec.Ledger.Serialization as SL
```
Comment (intricate) [inline ouroboros-consensus-shelley/test/Test/Consensus/Shelley/Ledger.hs:49]: This should come after the `Shelley.Spec.Ledger.Scripts` import.

File: ouroboros-consensus-shelley/test/Test/Consensus/Shelley/Ledger.hs [hunk 4]
```diff
@@ -46,6 +46,8 @@ import qualified Shelley.Spec.Ledger.LedgerState as SL
...
+import qualified Shelley.Spec.Ledger.Scripts as SL
```
Comment (intricate) [inline ouroboros-consensus-shelley/test/Test/Consensus/Shelley/Ledger.hs:50]:
```suggestion
```

This import already exists. See the line below it.

**Comment by redxaxder** [comment]
> We also want to use the annotation to make it cheap to return the size of the header, given a block, see
> 
> https://github.com/input-output-hk/ouroboros-network/blob/918b8d4ca2df20abfe8bbaf21fe0eb86d5b9a2fe/ouroboros-consensus-shelley/src/Ouroboros/Consensus/Shelley/Ledger/Block.hs#L203
> 
> . Can you implement that? Just to check that it works.

`serialize` uses the `ToCBOR` instance for the header. That will change to use the header's annotation once it is added, so this will become cheap without making changes here.

**Comment by mrBliss** [comment]
I'm pleasantly surprised about how *un*invasive this PR is :slightly_smiling_face:

File: cabal.project [hunk 5]
```diff
@@ -159,42 +159,42 @@ source-repository-package
 source-repository-package
   type: git
   location: https://github.com/input-output-hk/cardano-ledger-specs
-  tag: f1d5ddb25531b2512796d34e9cfd803f7af76567
+  tag: e7116d427a13c56b83a48d2785376dd8509f087e
   --sha256: 1nxqdi8vbx3mg1qdbvj7v6ngfjbgbf9i7yxc3ijac2mqb4fn5b4f
```
[the remark this answers is not in the document]
Comment (redxaxder) [reply cabal.project:163]: I'll update them once the shelley header changes are merged.

**Comment by mrBliss** [comment]
> I still have to have a proper look at how annotations work for Shelley, but I have some feedback already, see the inline comments.
> 
> We also want to use the annotation to make it cheap to return the size of the header, given a block, see
> 
> https://github.com/input-output-hk/ouroboros-network/blob/918b8d4ca2df20abfe8bbaf21fe0eb86d5b9a2fe/ouroboros-consensus-shelley/src/Ouroboros/Consensus/Shelley/Ledger/Block.hs#L203
> Can you implement that? Just to check that it works.

Can you update this comment: https://github.com/input-output-hk/ouroboros-network/blob/918b8d4ca2df20abfe8bbaf21fe0eb86d5b9a2fe/ouroboros-consensus-shelley/src/Ouroboros/Consensus/Shelley/Ledger/Block.hs#L201
to:
```
      -- The Shelley decoders use annotations, so this is cheap
```


---

Are annotations for transactions planned for the future?

**Comment by redxaxder** [comment]
> Are annotations for transactions planned for the future?

Yes. Those are up next.

**Comment by mrBliss** [comment]
Please squash before merging.

**Comment by mrBliss** [comment]
The Windows CI failure should be fixed by rebasing onto master (see #1937).

**Comment by redxaxder** [comment]
bors merge

**Pull request merged**